lib/ecies/electrum-ecies.js

```diff
@@ -0,0 +1,150 @@
+  this.opts = opts || {}
+  this.opts.ephemeralKey = true
+}
+
+ECIES.prototype.privateKey = function (privateKey) {
+  $.checkArgument(PrivateKey.isValid(privateKey), 'no private key provided')
+
+  this._privateKey = PrivateKey(privateKey.toString()) || null
+  this.opts.ephemeralKey = false
+
+  return this
+}
+
+ECIES.prototype.publicKey = function (publicKey) {
+  $.checkArgument(PublicKey.isValid(publicKey), 'no public key provided')
+
+  this._publicKey = PublicKey(publicKey.toString()) || null
+
+  return this
+}
+
+var cachedProperty = function (name, getter) {
+  var cachedName = '_' + name
+  Object.defineProperty(ECIES.prototype, name, {
+    configurable: false,
+    enumerable: true,
+    get: function () {
+      var value = this[cachedName]
+      if (!value) {
+        value = this[cachedName] = getter.apply(this)
+      }
+      return value
+    }
+  })
+}
+
+cachedProperty('Rbuf', function () {
+  return this._privateKey.publicKey.toDER(true)
+})
+
+cachedProperty('ivkEkM', function () {
+  var r = this._privateKey.bn
+  var KB = this._publicKey.point
+  var P = KB.mul(r)
+  var S = PublicKey(P)
+  var Sbuf = S.toBuffer()
+  return Hash.sha512(Sbuf)
+})
+
+cachedProperty('iv', function () {
+  return this.ivkEkM.slice(0, 16)
+})
+
+cachedProperty('kE', function () {
+  return this.ivkEkM.slice(16, 32)
+})
+
+cachedProperty('kM', function () {
+  return this.ivkEkM.slice(32, 64)
+})
+
+// Encrypts the message (String or Buffer).
+ECIES.prototype.encrypt = function (message, ivbuf) {
+  if (!Buffer.isBuffer(message)) message = Buffer.from(message)
+  var ciphertext = AESCBC.encrypt(message, this.kE, this.iv)
+  var encbuf
+  var BIE1 = Buffer.from('BIE1')
+  if (this.opts.noKey && !this.opts.ephemeralKey) {
+    encbuf = Buffer.concat([BIE1, ciphertext])
+  } else {
+    encbuf = Buffer.concat([BIE1, this.Rbuf, ciphertext])
+  }
+  var hmac = Hash.sha256hmac(encbuf, this.kM)
+  if (this.opts.shortTag) hmac = hmac.slice(0, 4)
+  return Buffer.concat([encbuf, hmac])
+}
+
+ECIES.prototype.decrypt = function (encbuf) {
+  $.checkArgument(encbuf)
+  var tagLength = 32
+  var offset = 4
+  if (this.opts.shortTag) {
+    tagLength = 4
+  }
+  var magic = encbuf.slice(0, 4)
+  if (!magic.equals(Buffer.from('BIE1'))) {
+    throw new errors.DecryptionError('Invalid Magic')
+  }
+  if (!this.opts.noKey) {
+    var pub
+    // BIE1 use compressed public key, length is always 33.
+    pub = encbuf.slice(4, 37)
+    this._publicKey = PublicKey.fromDER(pub)
+    offset = 37
+  }
+
+  var ciphertext = encbuf.slice(offset, encbuf.length - tagLength)
+  var hmac = encbuf.slice(encbuf.length - tagLength, encbuf.length)
+
+  var hmac2 = Hash.sha256hmac(encbuf.slice(0, encbuf.length - tagLength), this.kM)
+  if (this.opts.shortTag) hmac2 = hmac2.slice(0, 4)
+
+  if (!hmac.equals(hmac2)) {
+    throw new errors.DecryptionError('Invalid checksum')
+  }
+
+  return AESCBC.decrypt(ciphertext, this.kE, this.iv)
```

Comment:
here this should use .decryptCipherkey

Reply:
IV is used here.
So it may be decryptCipherkeyIV and IV, or decrypt, I think.